### ISSUE TITLE
Tracking promise expression for substitute

### ISSUE THREAD

**Issue by dgkf** [dgkf/R#124]
I was surprised to find that `substitute` is able to discover the promise expression from promises that have already been evaluated:

```r
local({
  test <- 1 + 2
  print(test)  # evalutes, test is now `3`
  substitute(test)
})
# [1] 3
# 1 + 2
```

From this, we can see that R must keep a record of what expressions were evaluated to produce results. This feels like a pretty substantial overhead for all variables.

I think this might be a helpful feature in situations where someone is doing some wild metaprogramming to modify function bodies, for instance injecting code at the start of a function call that might make use (and therefore `force` some variables) without affecting the functions internal use of metaprogramming calls like `substitute`. 

I'm a little unsure of whether this is something worth designing around.

Alternatively, in the special case of function calls, the call stack could be interrogated to find these expressions, thought cases like the one above in a `local()` would not work.

**Comment by dgkf** [comment]
False alarm, the original example I was using was a range `1:10` and I think due to that alt-rep it was having odd behavior:

```r
> local({
+   test <- 1:10
+   print(test)
+   substitute({ test })
+ })
 [1]  1  2  3  4  5  6  7  8  9 10
{
    1:10
}
```

This makes it _appear_ like it's forcing the variable but still recovering the code `1:10`, but I think this is just R displaying a range alt-rep that never gets materialized into another representation. Other cases don't show this behavior.

That said, even after being forced, values in ***functions*** can still find their originating code, so I think those need to be handled by looking at the call stack.

**Comment by dgkf** [comment]
Still might require tracking function environment promises:

```r
f <- function(x) {
  print(force(x))
  environment()
}

e <- f(1 + 2)
# [1] 3

substitute(x, e)
# 1 + 2
```

This means that the environment, not the call stack, keeps knowledge about the function call that produced the environment, even if promises produced in that environment have already been forced.

**Comment by dgkf** [comment]
And @DavisVaughan came to the rescue. Turns out promises _do not_ get replaced upon evaluation - instead they have a field that gets populated after being `force`ed. The object itself continues to be a promise.

And thank goodness for that! I was really hoping this wouldn't force me to revisit the internals of an environment.

**Comment by dgkf** [comment]
Now resolved in #125
